mem/buffer_pool.go

```diff
@@ -0,0 +1,156 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package mem
+
+import (
+	"sort"
+	"sync"
+	"sync/atomic"
+)
+
+// BufferPool is a pool of buffers that can be shared, resulting in
+// decreased memory allocation. Currently, in gRPC-go, it is only utilized
+// for parsing incoming messages.
+//
+// # Experimental
+//
+// Notice: This API is EXPERIMENTAL and may be changed or removed in a
+// later release.
+type BufferPool interface {
+	// Get returns a buffer with specified length from the pool.
+	//
+	// The returned byte slice may be not zero initialized.
+	Get(length int) []byte
+
+	// Put returns a buffer to the pool.
+	Put([]byte)
+}
+
+var defaultBufferPoolSizes = []int{
+	256,
+	4 << 10,  // 4KB
+	16 << 10, // 16KB (max HTTP/2 frame size used by gRPC)
+	64 << 10, // 64KB
+	1 << 20,  // 1MB
+}
+
+var defaultBufferPool = func() *atomic.Pointer[BufferPool] {
+	pool := NewBufferPool(defaultBufferPoolSizes...)
+	ptr := new(atomic.Pointer[BufferPool])
+	ptr.Store(&pool)
+	return ptr
+}()
+
+func DefaultBufferPool() BufferPool {
+	return *defaultBufferPool.Load()
+}
+
+func SetDefaultBufferPool(pool BufferPool) {
+	defaultBufferPool.Store(&pool)
+}
+
+func NewBufferPool(poolSizes ...int) BufferPool {
+	sort.Ints(poolSizes)
+	pools := make([]*sizedBufferPool, len(poolSizes))
+	for i, s := range poolSizes {
+		pools[i] = newBufferPool(s)
+	}
+	return &tieredBufferPool{
+		sizedPools:   pools,
+		fallbackPool: newBufferPool(0),
+	}
+}
+
+// tieredBufferPool implements the BufferPool interface with multiple tiers of
+// buffer pools for different sizes of buffers.
+type tieredBufferPool struct {
+	sizedPools   []*sizedBufferPool
+	fallbackPool *sizedBufferPool
+}
+
+func (p *tieredBufferPool) Get(size int) []byte {
+	return p.getPool(size).Get(size)
+}
+
+func (p *tieredBufferPool) Put(buf []byte) {
+	p.getPool(len(buf)).Put(buf)
+}
+
+func (p *tieredBufferPool) getPool(size int) *sizedBufferPool {
+	poolIdx := sort.Search(len(p.sizedPools), func(i int) bool {
+		return p.sizedPools[i].defaultSize >= size
+	})
+
+	if poolIdx == len(p.sizedPools) {
+		return p.fallbackPool
+	}
+
+	return p.sizedPools[poolIdx]
+}
+
+type sizedBufferPool struct {
+	pool        sync.Pool
+	defaultSize int
+}
+
+func (p *sizedBufferPool) Get(size int) []byte {
+	bs := *p.pool.Get().(*[]byte)
+
+	if cap(bs) < size {
+		p.pool.Put(&bs)
+
+		return make([]byte, size)
+	}
+
+	return bs[:size]
+}
+
+func (p *sizedBufferPool) Put(buf []byte) {
+	buf = buf[:cap(buf)]
+	// TODO: replace this loop with `clear`, though the compiler should be smart
+	// enough to optimize this.
+	for i := range buf {
+		buf[i] = 0
+	}
+	p.pool.Put(&buf)
```

Comment:
My reasoning for it was that there's no telling how long these buffers will stay in the pool. If something sensitive like passwords are in those buffers, it could be problematic to keep them around longer than they should be. Also, because `Buffer.ReadOnlyData` returns the raw slice, readers may index past the slice's `len` (if `len < cap`) and reveal the original data. Just seems like a good idea, especially now that 1.21 introduced `clear`

Reply:
Since we dropped support got Go1.20, we should be good to use `clear` here. Thanks.

Reply:
Yep, all done